File: autonav_ws/src/scr_core/src/performance.cpp
#include "scr_msgs/msg/performance_result.hpp"
#include "scr_core/performance.h"
#include "rclcpp/rclcpp.hpp"

namespace SCR
{
    Performance::Performance()
    {
    }

    Performance::Performance(const std::string &node, rclcpp::Publisher<scr_msgs::msg::PerformanceResult>::SharedPtr performancePublisher)
    {
        resultsPublisher = performancePublisher;
        this->node = node;
    }

    Performance::~Performance()
    {
    }

    void Performance::start(const std::string &name)
    {
        timers[name] = rclcpp::Clock().now().nanoseconds();
    }

    void Performance::end(const std::string &name)
    {
        auto now = rclcpp::Clock().now().nanoseconds();
        auto duration = now - timers[name];
        history[name].push_back(duration);

        if (history[name].size() > 500)
        {
            history[name].erase(history[name].begin());
        }

        if (minimum == 0 || duration < minimum)
        {
            minimum = duration;
        }

        if (maximum == 0 || duration > maximum)
        {
            maximum = duration;
        }

        publishData(name);
    }

    void Performance::publishData(const std::string &name)
    {
        auto results = scr_msgs::msg::PerformanceResult();
        results.key = name;
        results.average = 0;
        results.minimum = minimum / 1000000;
        results.maximum = maximum / 1000000;
        results.latest = history[name].back() / 1000000;
        results.latest_ns = history[name].back();
        results.node = node;

        auto size = history[name].size();
        for (auto &value : history[name])
        {
            results.average += value / 1000000;
        }
        results.average /= size;
        resultsPublisher->publish(results);
    }
}

File: autonav_ws/src/scr_core/scr_core/performance.py
from scr_msgs.msg import PerformanceResult
from rclpy.node import Node


class Performance:
    def __init__(self, node: Node):
        self.node = node
        self.publisher = node.create_publisher(PerformanceResult, "/scr/performance", 100)

        self.history = {}
        self.timers = {}
        self.minimum = 0
        self.maximum = 0

    def start(self, name):
        self.timers[name] = self.node.get_clock().now().nanoseconds

    def end(self, name):
        if name not in self.timers:
            return

        if name not in self.history:
            self.history[name] = []

        duration = self.node.get_clock().now().nanoseconds - self.timers[name]
        self.history[name].append(duration)

        if len(self.history[name]) > 500:
            self.history[name].pop(0)

        if duration < self.minimum or self.minimum == 0:
            self.minimum = duration

        if duration > self.maximum:
            self.maximum = duration

        self.publish(name)

    def publish(self, name):
        msg = PerformanceResult()
        msg.key = name
        msg.average = sum(self.history[name]) // len(self.history[name]) // 1000000
        msg.minimum = self.minimum // 1000000
        msg.maximum = self.maximum // 1000000
        msg.latest = self.history[name][-1] // 1000000
        msg.latest_ns = self.history[name][-1]
        msg.node = self.node.get_name()
        self.publisher.publish(msg)